orders-accounting/src/cache.rs

```diff
@@ -0,0 +1,258 @@
+    }
+
+    fn get_give_balance(&self, id: &OrderId) -> Result<Option<Amount>> {
+        let parent_supply = self.parent.get_give_balance(id).map_err(|_| Error::ViewFail)?;
+        let local_delta = self.data.give_balances.data().get(id).cloned();
+        combine_amount_delta(&parent_supply, &local_delta).map_err(Error::AccountingError)
+    }
+}
+
+impl<P: OrdersAccountingView> OrdersAccountingOperations for OrdersAccountingCache<P> {
+    fn create_order(&mut self, id: OrderId, data: OrderData) -> Result<OrdersAccountingUndo> {
+        if self.get_order_data(&id)?.is_some() {
+            return Err(Error::OrderAlreadyExists(id));
+        }
+
+        if self.get_ask_balance(&id)?.is_some() {
+            return Err(Error::OrderAlreadyExists(id));
+        }
+
+        let ask_value = data.ask().clone();
+        let give_value = data.give().clone();
+        let undo_data = self
+            .data
+            .order_data
+            .merge_delta_data_element(id, accounting::DataDelta::new(None, Some(data)))?;
+
+        self.data.ask_balances.add_unsigned(id, ask_value.amount())?;
+        self.data.give_balances.add_unsigned(id, give_value.amount())?;
+
+        Ok(OrdersAccountingUndo::CreateOrder(CreateOrderUndo {
+            id,
+            undo_data,
+            ask_balance: ask_value.amount(),
+            give_balance: give_value.amount(),
+        }))
+    }
+
+    fn withdraw_order(&mut self, id: OrderId) -> Result<OrdersAccountingUndo> {
+        let order_data = self
+            .get_order_data(&id)?
+            .ok_or(Error::AttemptedWithdrawNonexistingOrderData(id))?;
+        let ask_balance = self
+            .get_ask_balance(&id)?
+            .ok_or(Error::AttemptedWithdrawNonexistingAskBalance(id))?;
+        let give_balance = self
+            .get_give_balance(&id)?
+            .ok_or(Error::AttemptedWithdrawNonexistingGiveBalance(id))?;
+
+        let undo_data = self
+            .data
+            .order_data
+            .merge_delta_data_element(id, accounting::DataDelta::new(Some(order_data), None))?;
+
+        self.data.ask_balances.sub_unsigned(id, ask_balance)?;
+        self.data.give_balances.sub_unsigned(id, give_balance)?;
+
+        Ok(OrdersAccountingUndo::WithdrawOrder(WithdrawOrderUndo {
+            id,
+            undo_data,
+            ask_balance,
+            give_balance,
+        }))
+    }
+
+    fn fill_order(&mut self, id: OrderId, fill_value: OutputValue) -> Result<OrdersAccountingUndo> {
+        let filled_amount = calculate_fill_order(self, id, &fill_value)?;
+
+        let ask_balance = self.get_ask_balance(&id)?.ok_or(Error::OrderAskBalanceNotFound(id))?;
+        let give_balance =
+            self.get_give_balance(&id)?.ok_or(Error::OrderGiveBalanceNotFound(id))?;
+
+        // in case the order is completely filled it can be removed
+        let undo_data = if fill_value.amount() == ask_balance {
+            ensure!(
+                filled_amount == give_balance,
```

Comment:
yeah, I was so concentrated on making it correct that I missed the point that this doesn't make much practical sense. So we can't clean up orders automatically. Can it be solved with `sweep_orders` function in the wallet for now?

Reply:
I can imagine situations where it's impossible to fill an order. If you ask for a=10A for b=3B tokens, that means, the price is a/b=3, and you buy `(a/b)*3B=3*3B=9A`. The last 1A will never be filled no matter what you do. I don't know what valid ways there are to fill the order to the latter when there's such a difference, but we have to be careful with this as if this is done wrong, it can be abused.

Reply:
not exactly. I don't calculated price anywhere to avoid integer division. Instead the amounts exchanged are calculated which gives you (10A*3B)/3B so when buying 3b you would pay 10A. In case of filling it one buy one the actual "price" paid would be: 3A for the first and second token and the last token would cost you 4A to buy.